task/buildah/0.2/buildah.yaml

```diff
@@ -312,13 +318,16 @@ spec:
         VOLUME_MOUNTS="${VOLUME_MOUNTS} --volume ${mount_point}:${YUM_REPOS_D_TARGET}"
       fi
 
+      BUILD_DATE=$(date -u +'%Y-%m-%dT%H:%M:%S')
+      EPOCH=$(date -u --date="$BUILD_DATE" +'%s')
       LABELS=(
-        "--label" "build-date=$(date -u +'%Y-%m-%dT%H:%M:%S')"
+        "--label" "build-date=$BUILD_DATE"
         "--label" "architecture=$(uname -m)"
         "--label" "vcs-type=git"
       )
       [ -n "$COMMIT_SHA" ] && LABELS+=("--label" "vcs-ref=$COMMIT_SHA")
       [ -n "$IMAGE_EXPIRES_AFTER" ] && LABELS+=("--label" "quay.expires-after=$IMAGE_EXPIRES_AFTER")
+      [ "$EPOCH_AS_RELEASE_LABEL" == "true" ] && LABELS+=("--label" "release=$EPOCH")
```

Comment:
Not too happy about more redhat-isms in the buildah task. But at least this one is disabled by default.

Have you considered the alternative suggested in slack, where the buildah task would receive a generic `LABELS` param and users would need to pass the result of a generate-labels task as the value of this param?

```yaml
    - name: build-container
      params:
      - name: LABELS
        value:
          - $(tasks.generate-labels.results.LABELS[*])
```

It would require bigger changes on the users' part (adding a task vs. just setting a param for the buildah task) but could avoid yet another hardcoded redhat-ism

Reply:
As for how to approach a generate-labels task:

a) Hardcode redhat-specific labels, don't add it to the default pipeline, document in downstream docs how to use this task

b) Make it generic, document in downstream docs how to use it redhat-ically. Let's say something like this
```yaml
    - name: generate-labels
      params:
      - name: LABEL_TEMPLATES
        value:
          - "release=$epoch"
          - build-date=$build_date"
```
where the task defines a pre-determined set of placeholders and substitues them using whichever mechanism we'd like, e.g. `envsubst`

c) Let each team write their own

Reply:
Hm, I think I did consider it but I discarded it quickly in an emotional decision based on frustration that this release label keeps rearing its head; I had hoped we could be done with it. I'm more calm this morning, and so can slow down to think.

Hm, I want to isolate redhat-isms, but I also don't want to put a burden on new redhat users (who already have to work through enabling prefetch, hermetic, source containers, etc.)

Consider this combination:

* Select option (2). Make a generic task `generate-labels` as you described.
* But then, with kustomize, add that task with a redhat-ical configuration, to the default pipelines available in the redhat instances. We already have a need for this downstream-customization creeping up in the ecosystem preflight check task, which probably should not appear in the upstream default pipeline definitions.
* We can still document it as a how-to, but new downstream users in redhat would already start with the release label set as expected.

Is that the lowest-friction path?

Reply:
> But then, with kustomize, add that task with a redhat-ical configuration, to the default pipelines available in the redhat instances. We already have a need for this downstream-customization creeping up in the ecosystem preflight check task, which probably should not appear in the upstream default pipeline definitions.

As in, the customization would happen on the consumer side, down the stream from build-definitions? That sounds ideal. The consumer would have to build a modified Pipeline bundle, so it's not the most straightforward change if we want it fast, but it sounds like the right solution.

(And until then, we can document which labels to set in the downstream docs)

Reply:
Heh, side note. I think we're going to need a gettext step image. I don't think we have one already sitting around.

Reply:
Ah, true. Alternatively, python stdlib [string.Template](https://docs.python.org/3/library/string.html#template-strings) does something similar

```python
import collections
import os
import string
import sys

env = os.environ
# if you want envsubst behavior for nonexistent vars (replace with empty string)
# env = collections.defaultdict(str, env)

for arg in sys.argv[1:]:
    print(string.Template(arg).substitute(env))

```

Reply:
I added `gettext` to our yq image in https://github.com/konflux-ci/yq-container/pull/48

Reply:
OK, I updated the buildah task here to accept a new `ADDITIONAL_LABELS` param, and added a `generate-labels` task.

The idea is that you can use the `commit-timestamp` result from the `git-clone` task as the `SOURCE_DATE_EPOCH` param for the `generate-labels` task.

Eventually, we should drop the `build-date` label that's built into the buildah task and start expecting people to use generate-labels.